File: ios/Classes/EzvizView.swift
//
//  EzvizView.swift
//  flutter_ezviz
//
//  Created by 江鴻 on 2019/8/25.
//

import Foundation

class EzvizView : NSObject,FlutterPlatformView {
    
    private let player: EzvizPlayer
    private let methodChannel: FlutterMethodChannel
    
    init(messenger: FlutterBinaryMessenger, viewId: Int64, frame: CGRect) {
        player = EzvizPlayer(frame: frame)
        let methodChannelName = EzvizPlayerChannelMethods.methodChannelName + "_\(viewId)"
        methodChannel = FlutterMethodChannel.init(name: methodChannelName, binaryMessenger: messenger)
        super.init()
        methodChannel.setMethodCallHandler { [weak self](call, result) in
            self?.onHandle(call, result: result)
        }
    }
    
    deinit {
        player.playerRelease()
    }
    
    func view() -> UIView {
        return player
    }
    
    func onHandle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case EzvizPlayerChannelMethods.initPlayerByDevice:
            if let map = call.arguments as? Dictionary<String, Any> {
                let deviceSerial = map["deviceSerial"] as? String ?? ""
                let cameraNo = map["cameraNo"] as? Int ?? 0
                DispatchQueue.main.async {
                    self.player.initPlayer(deviceSerial: deviceSerial, cameraNo: cameraNo)
                }
            }
        case EzvizPlayerChannelMethods.initPlayerUrl:
            if let map = call.arguments as? Dictionary<String, Any> {
                let url = map["url"] as? String ?? ""
                DispatchQueue.main.async {
                    self.player.initPlayer(url: url)
                }
            }
        case EzvizPlayerChannelMethods.initPlayerByUser:
            if let map = call.arguments as? Dictionary<String, Any> {
                let userId = map["userId"] as? Int ?? 0
                let cameraNo = map["cameraNo"] as? Int ?? 0
                let streamType = map["streamType"] as? Int ?? 0
                DispatchQueue.main.async {
                    self.player.initPlayer(userId: userId, cameraNo: cameraNo, streamType: streamType)
                }
            }
        case EzvizPlayerChannelMethods.playerRelease:
            DispatchQueue.main.async {
                self.player.playerRelease()
            }
        case EzvizPlayerChannelMethods.setPlayVerifyCode:
            if let map = call.arguments as? Dictionary<String, Any> {
                let verifyCode = map["verifyCode"] as? String ?? ""
                DispatchQueue.main.async {
                    self.player.setPlayVerifyCode(verifyCode)
                }
            }
        case EzvizPlayerChannelMethods.startRealPlay:
            DispatchQueue.main.async {
                result(self.player.startRealPlay())
            }
        case EzvizPlayerChannelMethods.stopRealPlay:
            DispatchQueue.main.async {
                result(self.player.stopRealPlay())
            }
        case EzvizPlayerChannelMethods.startReplay:
            if let map = call.arguments as? Dictionary<String, Any> {
                let formatter = DateFormatter()
                formatter.dateFormat = "yyyyMMddHHmmss"
                let startTime = map["startTime"] as? String ?? ""
                let endTime = map["endTime"] as? String ?? ""
                let startDate = formatter.date(from: startTime) ?? Date()
                let endDate = formatter.date(from: endTime) ?? Date()
                DispatchQueue.main.async {
                    result(self.player.startReplay(startTime: startDate, endTime: endDate))
                }
            }
        case EzvizPlayerChannelMethods.stopReplay:
            DispatchQueue.main.async {
                result(self.player.stopReplay())
            }
        default:
            result(FlutterMethodNotImplemented)
        }
    }
}

File: android/src/main/kotlin/net/jayson/flutter_ezviz/EzvizView.kt
package net.jayson.flutter_ezviz

import android.content.Context
import android.view.View
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.PluginRegistry
import io.flutter.plugin.platform.PlatformView
import kotlinx.serialization.UnstableDefault
import kotlinx.serialization.json.Json
import java.text.SimpleDateFormat
import java.util.*

class EzvizFlutterPlayerView(context:Context, registrar: PluginRegistry.Registrar, id: Int) : PlatformView, MethodChannel.MethodCallHandler,EventChannel.StreamHandler,EzvizPlayerEventHandler {

    private val player: EzvizPlayerView
    private val methodChannel: MethodChannel
    private val eventChannel: EventChannel
    /// Native to flutter event
    private var eventSink: EventChannel.EventSink? = null

    init {
        player = EzvizPlayerView(context)
        player.eventHandler = this
        val methodChannelName = EzvizPlayerChannelMethods.methodChannelName + "_${id}"
        val eventChannelName = EzvizPlayerChannelEvents.eventChannelName + "_${id}"
        methodChannel = MethodChannel( registrar.messenger(),methodChannelName)
        eventChannel = EventChannel(registrar.messenger(),eventChannelName)
        eventChannel.setStreamHandler(this)
        methodChannel.setMethodCallHandler(this)
    }

    override fun getView(): View {
        return player
    }

    override fun dispose() {
        player.playRelease()
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when(call.method) {
            EzvizPlayerChannelMethods.initPlayerByDevice -> {
                val map = call.arguments as? Map<*, *>
                map?.let {
                    val deviceSerial = map["deviceSerial"] as? String ?: ""
                    val cameraNo = map["cameraNo"] as? Int ?: 0
                    player.initPlayer(deviceSerial,cameraNo)
                }
            }

            EzvizPlayerChannelMethods.initPlayerByUser -> {
                val map = call.arguments as? Map<*, *>
                map?.let {
                    val userId = map["userId"] as? Int ?: 0
                    val cameraNo = map["cameraNo"] as? Int ?: 0
                    val streamType = map["streamType"] as? Int ?: 0
                    player.initPlayer(userId, cameraNo, streamType)
                }
            }

            EzvizPlayerChannelMethods.initPlayerUrl -> {
                val map = call.arguments as? Map<*, *>
                map?.let {
                    val url = map["url"] as? String ?: ""
                    player.initPlayer(url)
                }
            }

            EzvizPlayerChannelMethods.playerRelease -> {
                player.playRelease()
            }

            EzvizPlayerChannelMethods.setPlayVerifyCode -> {
                val map = call.arguments as? Map<*, *>
                map?.let {
                    val verifyCode = map["verifyCode"] as? String ?: ""
                    player.setPlayVerifyCode(verifyCode)
                }
            }

            EzvizPlayerChannelMethods.startRealPlay -> {
                player.startRealPlay()
            }

            EzvizPlayerChannelMethods.stopRealPlay -> {
                player.stopRealPlay()
            }

            EzvizPlayerChannelMethods.startReplay -> {
                val map = call.arguments as? Map<*, *>
                map?.let {
                    val startTime = map["startTime"] as? String ?: ""
                    val endTime = map["endTime"] as? String ?: ""
                    val localDateFormat = SimpleDateFormat("yyyyMMddHHmmss", Locale.getDefault())
                    val startDate = localDateFormat.parse(startTime) ?: Date()
                    val endDate = localDateFormat.parse(endTime) ?: Date()
                    val startCalendar = Calendar.getInstance()
                    startCalendar.time = startDate
                    val endCalendar = Calendar.getInstance()
                    endCalendar.time = endDate
                    player.startPlayback(startCalendar,endCalendar)
                }
            }
            EzvizPlayerChannelMethods.stopReplay -> {
                player.stopPlayBack()
            }

            else -> {
                result.notImplemented()
            }
        }
    }

    override fun onListen(p0: Any?, p1: EventChannel.EventSink?) {
        this.eventSink = p1
    }

    override fun onCancel(p0: Any?) {
        this.eventSink = null
    }

    @UnstableDefault
    override fun onDispatchStatus(event: EzvizEventResult) {
        this.eventSink?.success(Json.stringify(EzvizEventResult.serializer(),event))
    }
}